### ISSUE TITLE
documentation is wrong about location of nodeVersion and npmVersion

### ISSUE THREAD

**Issue by fico-juliuskabugu** [eirslett/frontend-maven-plugin#294]
I had to put nodeVersion and npmVersion in the plugin configuration  because it was failing in the install-node-and-npm execution configuration. In the latter case, mvn complains that the parameters nodeVersion and npmVersion are missing or invalid.

**Comment by eirslett** [comment]
Ah, since the properties are defined inside the `<execution>` stuff instead of in the `<plugin>`/global parameters?
Would you care to send a quick PR that updates the documentation? It's possible to do inline editing of text files in GitHub, directly via the browser.

**Comment by demonfiddler** [comment]
I have just encountered the same problem. It's pretty inconvenient to have to share the same parameters between all executions of a single plugin element, it means that you can't execute the same goal more than once with different parameter values, or use multiple goals which use the same parameters names. I suggest that rather than updating the docs it would be rather more useful to update the code to support the de facto standard two-level Mojo configuration pattern ;-)

Example:
```xml
      <plugin>
        <groupId>com.github.eirslett</groupId>
        <artifactId>frontend-maven-plugin</artifactId>
        <version>0.0.26</version>
        <executions>
          <execution>
            <id>install-node</id>
            <goals>
              <goal>install-node-and-npm</goal>
            </goals>
            <!-- For reasons unknown, an execution-specific configuration is ignored. -->
            <!--configuration>
              <nodeVersion>v4.1.2</nodeVersion>
              <npmVersion>1.4.9</npmVersion>
            </configuration-->
          </execution>
          <execution>
            <id>install-tern</id>
            <goals>
              <goal>npm</goal>
            </goals>
            <!-- For reasons unknown, an execution-specific configuration is ignored. -->
            <!--configuration>
              <arguments>install tern</arguments>
              <workingDirectory>target/node</workingDirectory>
            </configuration-->
          </execution>
        </executions>
        <!-- For reasons unknown, configuration for all executions must be placed at the plugin level!!! -->
        <configuration>
          <nodeVersion>v4.1.2</nodeVersion>
          <npmVersion>1.4.9</npmVersion>
          <installDirectory>target</installDirectory>
          <arguments>install tern</arguments>
          <workingDirectory>target/node</workingDirectory>
        </configuration>
      </plugin>

**Comment by eirslett** [comment]
Do you have a use case where you want to run multiple versions of node/npm and multiple installation directories and working directories, within the same project/subproject? That certainly sounds like an edge case...

**Comment by demonfiddler** [comment]
Right this moment, no, after a lengthy debugging session I was able to get it working with the above configuration. It does make sense that the basic node/npm configuration be defined at the plug-in level but that wasn't my point.

In the above POM the <arguments>install tern</arguments> element does not belong at the plugin/configuration level since it relates only to plugin/execution[id = 'install-tern']. Right now there's no way for a single <plugin> element to contain more than one <execution> of npm, bower, jspm, grunt or gulp since they all use the same plugin/configuration/arguments element. To avoid collisions that element needs to be plugin/execution/configuration/arguments. That doesn't sound like an edge case, isn't this just a straightforward logical requirement?

**Comment by eirslett** [comment]
Of course. My impression was that one could specify node/npm versions etc. on the plugin level, and npm arguments on the execution level. Isn't that possible? [Like this?](https://github.com/eirslett/frontend-maven-plugin/blob/master/frontend-maven-plugin/src/it/custom-install-directory/pom.xml)

**Comment by demonfiddler** [comment]
Nope, splitting the configuration between the plugin and execution levels doesn't work. Per POM comments, if I move the plugin/configuration/arguments element to plugin/execution[id='install-tern']/configuration/arguments, npm fails with the ENOPACKAGEJSON error because the 'install tern' strings haven't been passed as arguments to npm.

**Comment by eirslett** [comment]
Have you tried moving tern into the "package.json" file, and using only "npm install"? Then, npm would install whatever is in the package.json file. That's how the example projects are setup, I haven't tried installing things directly via the command line yet.

**Comment by demonfiddler** [comment]
Yes, no doubt that could be made to work but it doesn't address the fundamental problem, which is that there is no way for a single plugin element to support multiple executions if two or more of those need distinct arguments.

**Comment by aksakalli** [comment]
sample project under `frontend-maven-plugin/src/it/example project/` should also be corrected.

**Comment by arunkjn** [comment]
this is completely broken after 1.3. Any updates on the issue?

**Comment by demonfiddler** [comment]
To be brutally honest chaps, isn't it blindingly obvious that something in this otherwise useful plug-in is simply broken from the configuration perspective?

From my experience of many other Maven plug-ins, it is an extremely common pattern to be able to place configuration at the either plugin level or the execution level and possibly even in both places. All we're asking is for frontend-maven-plugin to support this standard pattern. I don't think we should have to justify this expectation with uses cases or counter the claim that we're talking about some inconsequential edge case.

I don't have any insight into why this plug-in doesn't support this common pattern, I'd kind of assumed that the pattern was directly supported by Maven itself. Maybe it is, maybe it isn't, but it clearly is possible to achieve.

Please can't we just have a fix?

**Comment by eirslett** [comment]
Well here we are 8 years later, I still don't think this is fixed. I'm just going to close this issue, but if somebody is still bothered by this, or has the time to come up with a fix, that would be cool.
